### PR TITLE
Predict segmentation

### PR DESCRIPTION
**Pull request by dafna-r**
adding segmentation predict

### CODE REVIEW

File: src/super_gradients/training/processing/processing.py [hunk 1]
```diff
@@ -357,6 +371,162 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+    If more than one argument is given, the rescaling mode is determined by this order: scale_factor, then short_size,
+    then long_size.
+
+    :param scale_factor: Rescaling is done by multiplying input size by scale_factor:
+        out_size = (scale_factor * w, scale_factor * h)
+    :param short_size:  Rescaling is done by determining the scale factor by the ratio short_size / min(h, w).
+    :param long_size:   Rescaling is done by determining the scale factor by the ratio long_size / max(h, w).
+    """
+
+    def __init__(self, scale_factor: float, short_size: int, long_size: int):
```
Comment (Louis-Dupont) [inline src/super_gradients/training/processing/processing.py:424]: Let's add same default as the transform

File: src/super_gradients/training/processing/processing.py [hunk 2]
```diff
@@ -357,6 +371,162 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+            metadata.padding_coordinates.top : predictions.segmentation_map_shape[0] - metadata.padding_coordinates.bottom,
+            metadata.padding_coordinates.left : predictions.segmentation_map_shape[1] - metadata.padding_coordinates.right,
+        ]
+        predictions.segmentation_map = _rescale_image(predictions.segmentation_map.astype(np.uint8), target_shape=metadata.original_shape)
+        predictions.segmentation_map = predictions.segmentation_map.astype(np.int64)
+        return predictions
+
+
+@register_processing(Processings.SegmentationRescale)
+class SegmentationRescale(Processing, ABC):
```
Comment (Louis-Dupont) [inline src/super_gradients/training/processing/processing.py:412]: Why ABC? Is that a mistake ?

File: src/super_gradients/training/processing/processing.py [hunk 3]
```diff
@@ -357,6 +371,162 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
         return cropped_image, None
 
 
+@register_processing(Processings.SegResizeWithPadding)
+class SegResizeWithPadding(Processing, ABC):
```
Comment (Louis-Dupont) [inline src/super_gradients/training/processing/processing.py:375]: Why ABC? Is that a mistake ?

File: src/super_gradients/training/pipelines/pipelines.py [hunk 4]
```diff
@@ -196,7 +199,7 @@ def _generate_prediction_result_single_batch(self, images: Iterable[np.ndarray])
             postprocessed_predictions.append(prediction)
 
         # Yield results one by one
-        for image, prediction in zip(images, postprocessed_predictions):
+        for image, prediction, preprocessed_image in zip(images, postprocessed_predictions, preprocessed_images):
```
Comment (Louis-Dupont) [inline src/super_gradients/training/pipelines/pipelines.py:202]: That's a leftover right ? Because I dont see `preprocessed_image` being used

**Review by Louis-Dupont — DISMISSED** [review]
Looks Great!
I left some comments, mainly on style/convention.
I just have one more question; is there a reason why we don't make all segmentation models inherit from `SegmentationModule` ?

File: src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py [hunk 5]
```diff
@@ -174,6 +174,12 @@ def _get_pascal_labels(self) -> np.ndarray:
             ]
         )
 
+    def output_image_shape(self):
+        """
+        returns image shape when data set contains images of uniform shape.
+        """
+        return [512, 512]
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py]: fixed (036d2d4bb494e250a893a34edb873e0103f8c160 )

File: src/super_gradients/training/datasets/segmentation_datasets/coco_segmentation.py [hunk 6]
```diff
@@ -158,3 +158,9 @@ def _sub_dataset_creation(self, sub_dataset_image_ids_file_path) -> list:
         print("Number of images in sub-dataset: ", len(sub_dataset_image_ids))
         torch.save(sub_dataset_image_ids, sub_dataset_image_ids_file_path)
         return sub_dataset_image_ids
+
+    def output_image_shape(self):
+        """
+        returns image shape when data set contains images of uniform shape.
+        """
+        return [512, 512]
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/coco_segmentation.py]: fixed (036d2d4bb494e250a893a34edb873e0103f8c160 )

File: src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py [hunk 7]
```diff
@@ -206,3 +206,23 @@ def _transform_image_and_mask(self, image, mask) -> tuple:
...
+    def get_dataset_preprocessing_params(self):
+        """
+        Return any hardcoded preprocessing + adaptation for PIL.Image image reading (RGB).
+         image_processor as returned as a list of dicts to be resolved by processing factory.
+        :return:
+        """
+        pipeline = [Processings.ImagePermute()]
+        pipeline += [{Processings.NormalizeImage: {"mean": [0.485, 0.456, 0.406], "std": [0.229, 0.224, 0.225]}}]
+        if self.output_image_shape:
+            pipeline += [{Processings.SegResizeWithPadding: {"output_shape": self.output_image_shape, "pad_value": 0}}]
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py:223]: fixed (036d2d4bb494e250a893a34edb873e0103f8c160 )

File: src/super_gradients/training/datasets/segmentation_datasets/cityscape_segmentation.py [hunk 8]
```diff
@@ -149,6 +149,12 @@ def target_transform(target):
         out[out == 255] = CITYSCAPES_IGNORE_LABEL
         return out
 
+    def output_image_shape(self):
+        """
+        returns image shape when data set contains images of uniform shape.
+        """
+        return [1024, 2048]
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/cityscape_segmentation.py]: fixed (036d2d4bb494e250a893a34edb873e0103f8c160 )

File: src/super_gradients/training/processing/processing.py [hunk 9]
```diff
@@ -357,6 +371,162 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+
+@register_processing(Processings.SegmentationResize)
+class SegmentationResize(Processing, ABC):
+    """Resize image to given image dimensions.
+
+    :param output_h:    output shape will be (output_h, output_w)
+    :param output_w:    output shape will be (output_h, output_w)
+    """
+
+    def __init__(self, output_h: int, output_w: int):
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: fixed (b55fcd97f9cf830a80fe7a9ed00aec6999fb374a)

File: src/super_gradients/training/processing/processing.py (hunk 2 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: fixed (b55fcd97f9cf830a80fe7a9ed00aec6999fb374a)

File: src/super_gradients/training/processing/processing.py (hunk 3 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: fixed (b55fcd97f9cf830a80fe7a9ed00aec6999fb374a)

File: src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py [hunk 10]
```diff
@@ -206,3 +206,23 @@ def _transform_image_and_mask(self, image, mask) -> tuple:
         """
         transformed = self.transforms({"image": image, "mask": mask})
         return transformed["image"], transformed["mask"]
+
+    @property
+    def output_image_shape(self) -> Optional[Tuple[int, int]]:
+        return
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py]: done (d09e0247ce56bb72f415558a9f20b0327fc19d6b)

File: src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/segmentation_dataset.py:228]: done (8589cd21b5b58996d3fe976f3bf50bd18e9501f5)

File: src/super_gradients/training/processing/processing.py (hunk 1 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py:424]: not sure I understand your comment,
the transform and processing use the same logic/ order in choosing the scaling factor (both call "_compute_scale_factor" to determine the scaling factor)

File: src/super_gradients/training/pipelines/pipelines.py (hunk 4 above)
Comment (dafna-r) [reply src/super_gradients/training/pipelines/pipelines.py]: correct, fixed (1a3439dc1900e610f268a2dc75044fe2f1cb0be3)

File: src/super_gradients/training/models/segmentation_models/ppliteseg.py [hunk 11]
```diff
@@ -12,6 +13,12 @@
...
 from super_gradients.training.models.segmentation_models.context_modules import SPPM
+from super_gradients.training.pipelines.pipelines import SegmentationPipeline
+from super_gradients.common.decorators.factory_decorator import resolve_param
+from super_gradients.common.factories.processing_factory import ProcessingFactory
+from super_gradients.training.processing.processing import Processing
+from super_gradients.training.utils.predict import ImagesSegmentationPrediction
+from super_gradients.training.utils.media.image import ImageSource
 
 
 class UAFM(nn.Module):
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/models/segmentation_models/ppliteseg.py:24]: I'm not sure why, but most segmentation models don't inherit from SegmentationModule
I didn't make any changes to the models - this is how it was implemented before I started working on the predict

File: src/super_gradients/training/models/segmentation_models/segmentation_module.py [hunk 12]
```diff
@@ -59,3 +70,52 @@ def init_params(self):
...
+        )
+        return pipeline
+
+    def predict(self, images: ImageSource, batch_size: int = 32, fuse_model: bool = True) -> ImagesSegmentationPrediction:
+        """Predict an image or a list of images.
+        #
+        #     :param images:  Images to predict.
+        #     :param batch_size:  Maximum number of images to process at the same time.
+        #     :param fuse_model: If True, create a copy of the model, and fuse some of its layers to increase performance. This increases memory usage.
+        #"""
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/models/segmentation_models/segmentation_module.py]: yes, fixed (152f1751cceb04e860d4d6ace05718d1377410bb)

File: src/super_gradients/training/pipelines/pipelines.py [hunk 13]
```diff
@@ -442,3 +445,74 @@ def _combine_image_prediction_to_video(
...
+    :param image_processor:             Single image processor or a list of image processors for preprocessing and postprocessing the images.
+    :param device:                      The device on which the model will be run. If None, will run on current model device. Use "cuda" for GPU support.
+    :param fuse_model:                  If True, create a copy of the model, and fuse some of its layers to increase performance. This increases memory usage.
+    """
+
+    def __init__(
+        self,
+        model: SgModule,
+        class_names: List[str],
+        # post_prediction_callback, #post_prediction_callback: DetectionPostPredictionCallback,
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/pipelines/pipelines.py]: yes, fixed (0cbce7182119ca84417ae70d720f60b8c6d6c588)

File: src/super_gradients/training/pipelines/pipelines.py [hunk 14]
```diff
@@ -442,3 +445,74 @@ def _combine_image_prediction_to_video(
...
+        device: Optional[str] = None,
+        image_processor: Optional[Processing] = None,
+        fuse_model: bool = True,
+    ):
+        super().__init__(model=model, device=device, image_processor=image_processor, class_names=class_names, fuse_model=fuse_model)
+        # self.post_prediction_callback = post_prediction_callback
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/pipelines/pipelines.py]: yes, fixed (0cbce7182119ca84417ae70d720f60b8c6d6c588)

File: src/super_gradients/training/pipelines/pipelines.py [hunk 15]
```diff
@@ -442,3 +445,74 @@ def _combine_image_prediction_to_video(
...
+        """Decode the model output, by applying post prediction callback. This includes NMS.
+
+        :param model_output:    Direct output of the model, without any post-processing.
+        :param model_input:     Model input (i.e. images after preprocessing).
+        :return:                Predicted Bboxes.
+        """
+
+        if type(model_output) is tuple:
+            model_output = model_output(0)
+        # model_output = torch.nn.functional.softmax(model_output, dim=1)
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/pipelines/pipelines.py]: yes, fixed (0cbce7182119ca84417ae70d720f60b8c6d6c588)

File: src/super_gradients/training/processing/processing.py [hunk 16]
```diff
@@ -357,6 +371,162 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+        :param divisible_value:   the divisible value, new image size is an int multiplication of this number
+    =    :param fill_image:  Grey value to fill image padded background.
+    """
+
+    def __init__(self, divisible_value: int, fill_image: Union[int, Tuple, List]):
+        self.divisible_value = divisible_value
+        self.fill_image = fill_image
+
+    def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, DetectionPadToSizeMetadata]:
+        h, w = image.shape
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: fixed (700f99828fa3e1652819cb2759953397135f8760)

File: src/super_gradients/training/pipelines/pipelines.py [hunk 17]
```diff
@@ -442,3 +445,74 @@ def _combine_image_prediction_to_video(
...
+            images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Images")]
+
+        return ImagesSegmentationPrediction(_images_prediction_lst=images_predictions)
+
+    def _combine_image_prediction_to_video(
+        self, images_predictions: Iterable[ImageSegmentationPrediction], fps: float, n_images: Optional[int] = None
+    ) -> VideoDetectionPrediction:  # change to VideoSegmentationPrediction when implemented
+        raise NotImplementedError("This feature is not available for Segmentation task")
+        # images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
+        # return VideoSegmentationPrediction(_images_prediction_lst=images_predictions, fps=fps)
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/pipelines/pipelines.py]: added video segmentation prediction (eb7790444cccc4d0d0b106ca04d556f51054fb2b)

File: src/super_gradients/training/processing/processing.py [hunk 18]
```diff
@@ -519,6 +689,46 @@ def default_resnet_imagenet_processing_params() -> dict:
     return params
 
 
+def default_ppliteseg_cityscapes_processing_params() -> dict:
+    """Processing parameters commonly used for training ppliteseg on Cityscapes dataset."""
+    # image_processor = ComposeProcessing(
+    #     [Resize(size=256), CenterCrop(size=224), NormalizeImage(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]), StandardizeImage(), ImagePermute()]
+    # )
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: yes, fixed (bc529fa5360d4596451e8c14d177b98ff8a0163f)

File: src/super_gradients/examples/predict/segmentation_predict.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+from super_gradients.training import models
+
+# Note that currently only XXX, YYY and ZZZ are supported.
+# model = models.get(Models.PP_LITE_B_SEG, pretrained_weights="cityscapes")
+model = models.get(model_name=Models.PP_LITE_T_SEG75, arch_params={"use_aux_heads": False}, num_classes=19, pretrained_weights="cityscapes")
+
+IMAGES = [
+    "../../../../documentation/source/images/examples/countryside.jpg",
+    "../../../../documentation/source/images/examples/street_busy.jpg",
+    "https://cdn-attachments.timesofmalta.com/cc1eceadde40d2940bc5dd20692901371622153217-1301777007-4d978a6f-620x348.jpg",
```
Comment (BloodAxe) [inline src/super_gradients/examples/predict/segmentation_predict.py:11]: Let's not use URLs form the domains we do not control. @Louis-Dupont I recall a similar issue with some of the other PRs and if I'm not mistaken Ran Rubin uploaded some image to our S3 for this reason. Can you recall where it were and what was the url?

File: src/super_gradients/examples/predict/segmentation_predict.py [hunk 20]
```diff
@@ -0,0 +1,16 @@
+from super_gradients.common.object_names import Models
+from super_gradients.training import models
+
+# Note that currently only XXX, YYY and ZZZ are supported.
```
Comment (BloodAxe) [inline src/super_gradients/examples/predict/segmentation_predict.py:4]: Let's update XXX, YYY and ZZZ with up-to-date model names that are supported

File: src/super_gradients/training/processing/processing.py [hunk 21]
```diff
@@ -357,6 +371,160 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+        return processed_image, SegmentationRescaleWithPaddingMetadata(
+            original_shape=(height, width), scale_factor=scale_factor, padding_coordinates=padding_coordinates
+        )
+
+    def postprocess_predictions(self, predictions: SegmentationPrediction, metadata: SegmentationRescaleWithPaddingMetadata) -> SegmentationPrediction:
+        predictions.segmentation_map = predictions.segmentation_map[
+            metadata.padding_coordinates.top : predictions.segmentation_map_shape[0] - metadata.padding_coordinates.bottom,
+            metadata.padding_coordinates.left : predictions.segmentation_map_shape[1] - metadata.padding_coordinates.right,
+        ]
+        predictions.segmentation_map = _rescale_image(predictions.segmentation_map.astype(np.uint8), target_shape=metadata.original_shape)
```
Comment (BloodAxe) [inline src/super_gradients/training/processing/processing.py:406]: This is not the right way of resizing segmentation mask. 
Based on the code I assume you're working with hard labels here. The bilinear interpolation in this method would mess up labels. You should either use nearest interpolation when resizing semantic mask, of do a resizing in logits (before argmax) and then do argmax.

File: src/super_gradients/training/processing/processing.py [hunk 22]
```diff
@@ -357,6 +371,160 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+        return image, SegmentationResizeMetadata(original_shape=(height, width))
+
+    def postprocess_predictions(self, predictions: SegmentationPrediction, metadata: SegmentationResizeMetadata) -> SegmentationPrediction:
+        predictions.segmentation_map = _rescale_image(predictions.segmentation_map.astype(np.uint8), target_shape=metadata.original_shape)
+        predictions.segmentation_map = predictions.segmentation_map.astype(np.int64)
+        return predictions
+
+
+@register_processing(Processings.SegmentationPadShortToCropSize)
+class SegmentationPadShortToCropSize(Processing):
```
Comment (BloodAxe) [inline src/super_gradients/training/processing/processing.py:468]: Please add unit tests to all newly added classes. The DoD:
- Tests are passing for "normal" inputs.
- Tests are passing for 1x1 images
- Tests are handling a combination of images/params that are edge cases, like `SegmentationPadShortToCropSize(crop_size=(512,512)), input is 640,640`. Ideally behavior of transforms in such tests should be documented well.

File: src/super_gradients/training/processing/processing.py [hunk 23]
```diff
@@ -357,6 +371,160 @@ def preprocess_image(self, image: np.ndarray) -> Tuple[np.ndarray, None]:
...
+    If more than one argument is given, the rescaling mode is determined by this order: scale_factor, then short_size,
+    then long_size.
+
+    :param scale_factor: Rescaling is done by multiplying input size by scale_factor:
+        out_size = (scale_factor * w, scale_factor * h)
+    :param short_size:  Rescaling is done by determining the scale factor by the ratio short_size / min(h, w).
+    :param long_size:   Rescaling is done by determining the scale factor by the ratio long_size / max(h, w).
+    """
+
+    def __init__(self, scale_factor: float, short_size: int, long_size: int):
```
Comment (BloodAxe) [inline src/super_gradients/training/processing/processing.py:424]: I think it is a error prone design. IMO we should not confuse user by providing him a class, which behavior varies that significantly based on arguments.

To me it is much clearer to have two classes with a clear behavior and one parameter:
- SegmentationResizeLongestSizeTo(1024) - Would resize input image such that it's longest size is 1024
- SegmentationResizeShortestSizeTo(1024) - Similarly - would resize input image such that it's shortest size is 1024

Why two transforms and not three - because I cannot imagine any use case where i'd be willing to specify explicit scale factor over the absolute pixel values for resizing.

File: src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py [hunk 24]
```diff
@@ -174,6 +175,12 @@ def _get_pascal_labels(self) -> np.ndarray:
             ]
         )
 
+    def output_image_shape(self) -> Tuple[int, int]:
```
Comment (BloodAxe) [inline src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py:178]: I'm not quite following why we introduce functions with hard-coded return values here?

**Review by BloodAxe — CHANGES_REQUESTED** [review]
See my inline comments

File: src/super_gradients/training/processing/processing.py (hunk 21 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py:406]: Hi,
the predict should output the prediction mask (meaning classes) so the pre-processing can only work on the labels
you are right about the interpolation, we had an internal discussion about it before i committed the changes and the decision was to do it this way
@shaydeci @Louis-Dupont should i make any changes here?

File: src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py (hunk 24 above)
Comment (dafna-r) [reply src/super_gradients/training/datasets/segmentation_datasets/pascal_voc_segmentation.py:178]: most segmentation recipes have no transforms (manning no processing as well) - there's a hidden assumption that all inputs are of the same size.  
The prediction inputs might be with different sizes and we need to be able to run segmentation prediction on them. in addition the predict input size might be different from the input size the model trained on.
with this property we make sure to resize all prediction-inputs to the same size as the training data

File: src/super_gradients/training/utils/visualization/segmentation.py [hunk 25]
```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import torch
+from torchvision.utils import draw_segmentation_masks
+from typing import List, Optional, Tuple, Union, Set
+from super_gradients.training.utils.segmentation_utils import to_one_hot
+from data_gradients.visualize.detection.detection_legend import draw_legend_on_canvas
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/utils/visualization/segmentation.py]: draw_legend_on_canvas was added to visualization utils (super_gradients.training.utils.visualization.utils)
where all prediction visualization is 
(de6b6dbc22e47bb56cd5706c7786cb07442103a4)

File: src/super_gradients/training/processing/processing.py (hunk 22 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py:468]: done (2dc024f345594194a221854c69e2525307c9a516)

please note that this transform doesn't crop the image to into "crop_size", it only "enlarges" the image if needed (by padding)
if "crop_size" is larger than image size - the image will be padded into "crop size"
if "crop size" is smaller than image size - no change will be made  
the equivalent processing behaves in the same way

@Louis-Dupont @shaydeci - FYI

File: src/super_gradients/training/processing/processing.py (hunk 1 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: done (88b3d0af06a3118aa4d33e2aa1e0ecbfef19820b)

File: src/super_gradients/training/processing/processing.py (hunk 21 above)
Comment (Louis-Dupont) [reply src/super_gradients/training/processing/processing.py:406]: @dafna-r You can add am optional parameter `interpolation` to `_rescale_image` which by default would be `cv2.INTER_LINEAR`, and in your case you set it to `cv2. INTER_NEAREST `

Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: done (e3559e2d3cf4bb28359f6bce80268e8bf18e77ed)

File: src/super_gradients/examples/predict/segmentation_predict.py (hunk 19 above)
Comment (dafna-r) [reply src/super_gradients/examples/predict/segmentation_predict.py:10]: this is the same as all "predict" examples
if changes are made it should be done to all prediction examples

File: src/super_gradients/training/processing/processing.py (hunk 22 above)
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py:475]: Louis opened a ticket regarding transform/processing name (called crop but doesn't crop)

File: src/super_gradients/examples/predict/segmentation_predict.py (hunk 19 above)
Comment (BloodAxe) [reply src/super_gradients/examples/predict/segmentation_predict.py:10]: The fact that we have something wrong (Wrong url for detection predict) does not mean we should introduce new wrong things.
Please change `https://cdn-attachments.timesofmalta.com/cc1eceadde40d2940bc5dd20692901371622153217-1301777007-4d978a6f-620x348.jpg` to `https://deci-pretrained-models.s3.amazonaws.com/sample_images/beatles-abbeyroad.jpg`

File: src/super_gradients/training/models/segmentation_models/segmentation_module.py [hunk 26]
```diff
@@ -59,3 +70,51 @@ def init_params(self):
...
+
+        pipeline = SegmentationPipeline(
+            model=self,
+            image_processor=self._image_processor,
+            class_names=self._class_names,
+            fuse_model=fuse_model,
+        )
+        return pipeline
+
+    def predict(self, images: ImageSource, batch_size: int = 32, fuse_model: bool = True) -> ImagesSegmentationPrediction:
```
Comment (BloodAxe) [inline src/super_gradients/training/models/segmentation_models/segmentation_module.py:106]: Please add `HasPredict` mixin to inheritance: SegmentationModule(SgModule, ABC, HasPredict):

File: tests/unit_tests/segmentation_processing_PadShortToCropSize_test.py [hunk 27]
```diff
@@ -0,0 +1,82 @@
+import unittest
+import numpy as np
+
+from super_gradients.training.processing import SegmentationPadShortToCropSize
+from super_gradients.training.utils.predict.predictions import SegmentationPrediction
+
+
+class SegmentationPadShortToCropSizeTest(unittest.TestCase):
```
Comment (BloodAxe) [inline tests/unit_tests/segmentation_processing_PadShortToCropSize_test.py:8]: Please also add a test case for segmentation predict here https://github.com/Deci-AI/super-gradients/blob/master/tests/unit_tests/test_predict.py

File: src/super_gradients/training/models/segmentation_models/segmentation_module.py (hunk 26 above)
Comment (dafna-r) [reply src/super_gradients/training/models/segmentation_models/segmentation_module.py:107]: done (4df033eecccf78c3cd10dfe89771242bd3c0abcf)

File: tests/unit_tests/segmentation_processing_PadShortToCropSize_test.py (hunk 27 above)
Comment (dafna-r) [reply tests/unit_tests/segmentation_processing_PadShortToCropSize_test.py:8]: done (4df033eecccf78c3cd10dfe89771242bd3c0abcf)

File: src/super_gradients/training/processing/processing.py [hunk 28]
```diff
@@ -537,4 +749,7 @@ def get_pretrained_processing_params(model_name: str, pretrained_weights: str) -
     if pretrained_weights == "imagenet" and model_name == "resnet18":
         return default_resnet_imagenet_processing_params()
 
+    if pretrained_weights == "cityscapes" and model_name == "pp_lite_t_seg75":
```
[the remark this answers is not in the document]
Comment (dafna-r) [reply src/super_gradients/training/processing/processing.py]: added a few additional defaults (4df033eecccf78c3cd10dfe89771242bd3c0abcf)

**Comment by BloodAxe** [comment]
Replaced by https://github.com/Deci-AI/super-gradients/pull/1366
